vpr/src/route/route_timing.cpp

```diff
@@ -2004,38 +2003,100 @@ bool should_setup_lower_bound_connection_delays(int itry, const t_router_opts& /
+
+    // Worst negative hold slack is prioritized since getting this as close to zero as possible increases the odds of the routing still working on a physical device
+    //  this is because there are margins of error on timing calculations
+    if (timing_info->hold_worst_negative_slack() > best_routing_metrics.sWNS) {
+        return 1;
+    } else if (timing_info->hold_worst_negative_slack() < best_routing_metrics.sWNS) {
+        return -1;
+    }
+
+    if (timing_info->hold_total_negative_slack() > best_routing_metrics.sTNS) {
+        return 1;
+    } else if (timing_info->hold_total_negative_slack() < best_routing_metrics.sTNS) {
+        return -1;
+    }
+
+    // Report a tie
+    return 0;
+}
+
 static bool is_better_quality_routing(const vtr::vector<ClusterNetId, t_traceback>& best_routing,
                                       const RoutingMetrics& best_routing_metrics,
                                       const WirelengthInfo& wirelength_info,
-                                      std::shared_ptr<const SetupHoldTimingInfo> timing_info) {
+                                      std::shared_ptr<const SetupHoldTimingInfo> timing_info,
+                                      const t_router_opts& router_opts) {
     if (best_routing.empty()) {
         return true; //First legal routing
     }
 
-    //Rank first based on sWNS, followed by other timing metrics
     if (timing_info) {
-        if (timing_info->setup_worst_negative_slack() > best_routing_metrics.sWNS) {
-            return true;
-        } else if (timing_info->setup_worst_negative_slack() < best_routing_metrics.sWNS) {
-            return false;
-        }
+        if (router_opts.routing_budgets_algorithm != YOYO) {
+            // If RCV is disabled prioritize setup slack improvements
+            int setup_comparison = compare_setup(best_routing_metrics, timing_info);
+
+            if (setup_comparison > 0) {
+                // Setup has improved
+                return true;
+            } else if (setup_comparison < 0) {
+                // Setup has degraded
+                return false;
+            }
+            // Setup slack tied
+        } else {
+            // If RCV is enabled prioritize hold over setup, but still return true if setup slack improves
 
-        if (timing_info->setup_total_negative_slack() > best_routing_metrics.sTNS) {
-            return true;
-        } else if (timing_info->setup_total_negative_slack() < best_routing_metrics.sTNS) {
-            return false;
-        }
+            // Compare hold of current routing to best routing
+            int hold_comparison = compare_hold(best_routing_metrics, timing_info);
 
-        if (timing_info->hold_worst_negative_slack() > best_routing_metrics.hWNS) {
-            return true;
-        } else if (timing_info->hold_worst_negative_slack() > best_routing_metrics.hWNS) {
-            return false;
-        }
+            if (hold_comparison > 0) {
```

Comment:
This code could be shortened by checking if YOYO is enabled, and if so, checking hold and returning whatever is better, or doing nothing in a tie.  Then check setup (no matter what), then wirelength.

Reply:
You're right, I should have seen this! Will fix thanks.